export_logic.go

```diff
@@ -0,0 +1,366 @@
+					return err
+				}
+			}
+
+			// Export may have multiple branches with the same parent.
+			// We can't handle that currently so abort.
+			for _, v := range e.Repositories {
+				commits := map[string]string{}
+				for tag, commit := range *v {
+					commits[commit] = tag
+				}
+				if len(commits) > 1 {
+					return errorMultipleBranchesSameParent
+				}
+			}
+
+			continue
+		}
+
+		uuidPart := nameParts[0]
+		fileName := nameParts[1]
+		if e.Layers[uuidPart] == nil {
+			e.Layers[uuidPart] = &layer{}
+		}
+
+		switch fileName {
+		case "json":
+			e.Layers[uuidPart].JSONHeader = header
+			if err = json.NewDecoder(tarReader).Decode(&e.Layers[uuidPart].LayerConfig); err != nil {
+				return err
+			}
+		case "layer.tar":
+			layerReader := tar.NewReader(tarReader)
+			for {
+				fileHeader, err := layerReader.Next()
+				if err != nil {
+					if err == io.EOF {
+						break
+					}
+					return err
+				}
+				filePath := nameWithoutWhiteoutPrefix(fileHeader.Name)
+				if e.fileToLayers[filePath] == nil {
+					e.fileToLayers[filePath] = []fileLoc{}
+				}
+				foundWhiteout := isWhiteout(fileHeader.Name)
+				e.fileToLayers[filePath] = append(e.fileToLayers[filePath], fileLoc{
+					uuid:     uuidPart,
+					whiteout: foundWhiteout,
+				})
+
+				if foundWhiteout {
+					e.whiteouts = append(e.whiteouts, whiteoutFile{
+						prefix: filePath,
+						uuid:   uuidPart,
+					})
+				}
+			}
+		}
+	}
+
+	e.start = e.FirstSquash()
+
+	// Can't find a previously squashed layer, default to root
+	if e.start == nil {
+		e.start = e.Root()
+	}
+
+	if e.start == nil {
+		return errorNoFROM
+	}
+
+	// TODO: optimize creation of ordered list - currently n^2, can be n
+	index := 0
+	current := e.start
+	orderMap := map[string]int{}
+	for {
+		orderMap[current.LayerConfig.ID] = index
+		index++
+		current = e.ChildOf(current.LayerConfig.ID)
+		if current == nil {
+			break
+		}
+	}
+
+	for path, fileLocs := range e.fileToLayers {
+		if len(fileLocs) > 0 {
+			greatest := fileLocs[0]
+			for _, loc := range fileLocs {
+				if orderMap[loc.uuid] > orderMap[greatest.uuid] {
+					greatest = loc
+				}
+			}
+			if e.layerToFiles[greatest.uuid] == nil {
+				e.layerToFiles[greatest.uuid] = map[string]bool{}
+			}
+
+			// if name matches whiteout prefix and the whiteout file is found
+			// in a layer that is >= greatest.uuid, skip
+			uuidContainingWhiteout, matches := matchesWhiteout(path, e.whiteouts)
+			if (matches && orderMap[uuidContainingWhiteout] >= orderMap[greatest.uuid]) || greatest.whiteout {
+				delete(e.layerToFiles[greatest.uuid], path)
+			} else {
+				e.layerToFiles[greatest.uuid][path] = true
+			}
+		}
+	}
+
+	return nil
+}
+
+func (e *export) SquashLayers(into, from *layer, instream io.Reader, outstream io.Writer) (imageID string, err error) {
+	tempfile, err := ioutil.TempFile("", "libsquash")
+	if err != nil {
+		return "", err
+	}
+
+	defer func() {
+		_ = tempfile.Close()
```

Comment:
This is a personal preference of mine. I like to indicate that an `error` is being throw away - should it cause a problem, this at least makes it easier to find